grant-project/zcash-grant-system
# Proposal pagination & 404 fix

Closes #122. Part of #121.

### What this does
- pagination 'system'
- pagination shape:
```js
{
  filters: [...strings...], // may set this
  search: 'string', // may set this
  sort: 'string', // may set this
  page: 1, // may set this
  pageSize: 9,
  total: 100,
  items: [{...}, {...}, ..., {...}]
}
```
- backend interprets filter strings (using prefixes for Proposal) for sqlalchemy
- backend interprets sort string for sqlalchemy
- frontend sorting, filtering & searching (by title) working
- frontend redux change: `app.proposal.proposals` => `app.proposal.page`(list)
- frontend redux change: `app.proposal.proposals.find(...)` => `app.proposal.detail` (detail)
- admin store change: `store.proposals` => `store.proposals.page` (list)
- admin moved filter state off of URL and into `store.proposals.page.*`
- admin: added sorting, search & pagination
- fixes forever spinning when a bad proposalId is provided to proposal detail view #122 

### Thoughts
- this doesn't implement filtering/sorting on funding state, `contributed` and `funded` fields are hybrid properties so they need special care to get working with sql sorting and filtering, needs more [research](https://docs.sqlalchemy.org/en/latest/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior)
- frontend & pagination state is not cleared when browsing away from proposals list, we might want to clear it in certain situations

### Test
1. browse to a non-existing proposal `/proposals/999`, should get a nice 404
1. you may want to use the command ` flask create-proposals 100` this will create a bunch of uninspired garbage proposals in your DB giving to multiple pages to browse (you are left to your own devices to clean up the mess!)
1. browse to proposal list view, try all the things - search (1s debounce), filters, sorting, (stages not supported yet)
1. with an active search, browse away and back, the search term should populate the search INPUT
1. go to admin and play around with proposals - search (must hit enter), filter, sort
1. on admin, visit `/proposals?status=PENDING`, it should populate the correct filter and remove `?status=PENDING` from URL

## Screen
![pagination-demo](https://user-images.githubusercontent.com/11430954/52234759-09683b80-2888-11e9-801f-a4235f287822.gif)

![screen shot 2019-02-04 at 2 16 52 pm](https://user-images.githubusercontent.com/11430954/52234768-0f5e1c80-2888-11e9-8eac-06b5b2f0228f.png)
